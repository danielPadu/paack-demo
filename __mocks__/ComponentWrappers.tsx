import {NavigationContext} from '@react-navigation/native';
import React from 'react';
import {Provider} from 'react-redux/src';
import {PersistGate} from 'redux-persist/integration/react';
import {persistor} from '../appFiles/appStore/store';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import {defaultReducer} from '../appFiles/appStore/reducers/default';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
export const mockedStore = mockStore(defaultReducer);

export const navContext = {
  isFocused: () => true,
  // addListener returns an unscubscribe function.
  addListener: jest.fn(() => jest.fn()),
  navigate: Function.prototype,
  setParams: Function.prototype,
  dispatch: Function.prototype,
  getParam: (param: {deliveryId?: string} & any, defaultValue: any) => {
    return defaultValue;
  },
};

export const ComponentWrappers = (props: {
  children: React.ReactNode;
  redux_store: any;
}) => {
  const {children, redux_store} = props;

  return (
    <>
      <Provider store={redux_store ?? mockedStore}>
        <PersistGate loading={null} persistor={persistor}>
          <NavigationContext.Provider value={navContext as any}>
            {children}
          </NavigationContext.Provider>
        </PersistGate>
      </Provider>
    </>
  );
};
