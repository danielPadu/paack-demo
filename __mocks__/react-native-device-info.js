export default {
  getModel: jest.fn(),
  getVersion: jest.fn(),
  getBuildNumber: jest.fn(),
  getBundleId: jest.fn(),
  hasNotch: jest.fn(),
};
