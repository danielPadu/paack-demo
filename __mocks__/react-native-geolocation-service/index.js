export default {
  addListener: jest.fn(),
  getCurrentPosition: () => ({
    coords: {
      accuracy: 0,
      altitude: 0,
      altitudeAccuracy: 0,
      heading: 0,
      latitude: 0,
      longitude: 0,
      speed: 0,
    },
  }),
  removeListeners: jest.fn(),
  requestAuthorization: jest.fn(),
  setConfiguration: jest.fn(),
  startObserving: jest.fn(),
  stopObserving: jest.fn(),
};
