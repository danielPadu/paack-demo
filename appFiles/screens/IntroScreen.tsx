import {useNetInfo} from '@react-native-community/netinfo';
import {useIsFocused} from '@react-navigation/native';
import React, {useCallback, useEffect, useLayoutEffect, useState} from 'react';
import {SafeAreaView, ScrollView, StyleSheet, View} from 'react-native';
import {Button} from 'react-native-ui-lib';
import {useDispatch} from 'react-redux';
import {infoModalOpen} from '../appStore/actions/modalActions';
import useLocation from '../hooks/useLocation';
import {navigate} from '../navigation/RootNavigation';
import {LocationCard} from '../UI/components';
import {activityIndicator} from '../UI/components/buttonsActivityIndicator';
import {isIOS, log, NetworkUtils} from '../UI/utils';

const IntroScreen = () => {
  const netInfo = useNetInfo();
  const isFocused = useIsFocused();
  const dispatch = useDispatch();
  const [checkingPermissions, setCheckingPermissions] = useState(false);
  const [showLocation, setShowLocation] = useState(false);
  const {getLocation} = useLocation();
  useLayoutEffect(() => {
    if (isFocused) {
      NetworkUtils()?.then((res: any) => {
        log({'NetworkUtils: ': res});
        if (res.isInternetReachable === false) {
          dispatch(
            infoModalOpen({
              title: `Connectivity error ${
                netInfo.type !== 'none'
                  ? 'on ' + netInfo.type + ' connection'
                  : ''
              } `,
              description:
                'Your device can not acces Paack server. Check network connection and try again',
            }),
          );
        }
      });
    }
  }, [dispatch, isFocused, netInfo]);
  useEffect(() => {
    onCheckPermissionsPress();
  }, []);
  const onCheckPermissionsPress = () => {
    setCheckingPermissions(true);
    getLocation()
      .then(result => {
        if (result === 'granted') {
          navigate('DeliveriesScreen');
        }
      })
      .catch(result => {
        if (result === 'blocked') {
          dispatch(
            infoModalOpen({
              title: 'No permissions granted',
              description:
                'Your device did not granted accessing location. Please review Paack permissions in ',
            }),
          );
        }
      })
      .finally(() => setCheckingPermissions(false));
  };
  const {wrapperStyle, containerStyle, buttonsContainer} = styles;
  return (
    <SafeAreaView style={wrapperStyle}>
      <View style={containerStyle}>
        <ScrollView
          showsVerticalScrollIndicator={false}
          nestedScrollEnabled={false}>
          <View>
            <View style={buttonsContainer}>
              <Button
                label={`${showLocation ? 'Hide' : 'Show'} current location`}
                onPress={() => setShowLocation(c => !c)}
                bg-secondaryColor
                square
                marginH-10
              />
              <Button
                testID={'checkPermissionsButton'}
                label={
                  checkingPermissions ? 'Processing' : 'Navigate to deliveries'
                }
                disabled={checkingPermissions}
                onPress={onCheckPermissionsPress}
                iconOnRight
                iconSource={useCallback(
                  () => activityIndicator(checkingPermissions),
                  [checkingPermissions],
                )}
                bg-primaryColor
                square
              />
            </View>
            {showLocation && <LocationCard />}
          </View>
        </ScrollView>
      </View>
    </SafeAreaView>
  );
};
const styles = StyleSheet.create({
  wrapperStyle: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'space-between',
    margin: 0,
    backgroundColor: 'white',
    color: 'black',
  },

  containerStyle: {
    flex: 1,
    width: '100%',
    justifyContent: 'space-between',
    paddingTop: isIOS ? 10 : 100,
  },

  buttonsContainer: {
    margin: 0,
    marginTop: 10,
    padding: 0,
    marginHorizontal: 15,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  checkPermissionsButtonLoadingStyle: {marginLeft: 10},

  underlined: {
    textDecorationLine: 'underline',
    color: 'black',
    fontSize: 18,
    fontWeight: '400',
  },
  lightTextStyle: {color: 'darkgray', fontSize: 15},
});
export default IntroScreen;
