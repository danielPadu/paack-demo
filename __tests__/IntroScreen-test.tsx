import React from 'react';
import {fireEvent, waitFor} from 'react-native-testing-library';
import renderer, {
  act,
  ReactTestRenderer,
  ReactTestRendererJSON,
} from 'react-test-renderer';
import IntroScreen from '../appFiles/screens/IntroScreen';
import {ComponentWrappers, mockedStore} from '../__mocks__/ComponentWrappers';
let store;
let component:
  | ReactTestRenderer
  | ReactTestRenderer[]
  | ReactTestRendererJSON
  | ReactTestRendererJSON[]
  | null;

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: () => ({app_allow_geolocation: true, last_location: {}}),
  useDispatch: () => jest.fn(),
}));

describe('checking IntroScreen component', () => {
  it('IntroScreen renders ', async () => {
    await act(async () => {
      await beforeEach(() => {
        store = mockedStore;

        component = renderer.create(
          <ComponentWrappers redux_store={store}>
            <IntroScreen />
          </ComponentWrappers>,
        );
      });
      await afterEach(() => {
        // cleanup on exiting
        component?.remove?.();
        component = null;
      });
      expect(component).toMatchSnapshot();
    });
  });

  test('IntroScreen inner functionallity tests', async () => {
    await act(async () => {
      await beforeEach(async () => {
        store = mockedStore;

        component = renderer.create(
          <ComponentWrappers redux_store={store}>
            <IntroScreen />
          </ComponentWrappers>,
        );
        const {getByTestId} = component;
        const mockedHandler = jest.fn();
        const button = getByTestId('checkPermissionsButton');
        button.props.onPress = mockedHandler;
        fireEvent.press(button);
        await waitFor(() => expect(mockedHandler).toHaveBeenCalled());
        await waitFor(() => expect(component).toHaveProperty('netinfo'));
        await waitFor(() => expect(component.netinfo.exists()).toBe(true));
      });
      await afterEach(() => {
        // cleanup on exiting
        component?.remove?.();
        component = null;
      });
    });
  });
});
