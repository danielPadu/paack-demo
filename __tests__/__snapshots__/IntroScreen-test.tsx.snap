// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checking IntroScreen component IntroScreen renders  1`] = `undefined`;
